Create BinaryFileReader.py

anjalizope opened the pull request:
Added code for binary file reader

anjalizope committed:
Create BinaryFileReader.py

Added code for binary file reader

mar2194 commented:
Congrats Anjali!

ryan-brazeal-ufl reviewed (changes requested):
Some ideas I had for new OPL features built using Anjali’s contribution (look up the function/method based on these code snippets):

''' reader from importing LVX point and IMU data into Python3
points - numpy.ndarray
imu - list of lists
datatype - integer (0-5)
**points, imu, datatype = opl.LVXreader("horizon_f6.10.lvx")**
'''
''' file conversion support from LVX to CSV and LAS (just like Livox Viewer, which would be a great addition)
**opl.LVX2CSV(“mid100_f3.3.7.lvx", “test1.csv")
opl.LVX2LAS("mid40_f3.3.6.lvx", "test2.las”)**

The modifications I am proposing for BinaryFIleReader.py include:
- ability to suppress print statements
- added datatype parameter information being returned by the BinaryReaders object so the user knows if its Cartesian or Spherical data
- dual return mode data from a Horizon (and Tele-15) sensor records all single returns as well as dual returns per laser pulse
- returned point data units are either meters or decimal degrees
- point data includes return number attribute for every point (constant 1 for single return data types should also be included)
- standard order for returned point data (Cartesian and Spherical) - [X, Y, Z, reflectivity, returnNum] or [Distance, Zenith, Azimuth, reflectivity, returnNum]
- if sensor uses 1 byte ’tag information’ its integer equivalent is recorded after returnNum
- support for multi-return dataset from MID-40/100 via the special firmwares

I would propose that the following functionalities also be implemented within OPL (not currently included within the attached openpylivox.py file)
- include a precise timestamp (requires simple calculation) for each point and IMU record
- HERE’S A BIG ONE … include support for LVX file creation (i.e., a writer class) directly from OPL binary data.
* In case you haven’t done the deep dive into OPL, yet… 
   - currently OPL writes the point and IMU data it collects from a Livox sensor in near real-time to a simple sequential binary file format (creates a .bin extension file)
   - currently OPL binary data can be converted to CSV and LAS
   - support writing/converting OPL binary data to the LVX file format, which would allow for the data to be viewed, calibrated, played-back, etc. using the Livox-Viewer application.

anjalizope committed:
Update BinaryReaders.py with modifications

Modifications include:
- ability to suppress print statements
- added datatype parameter information being returned by the BinaryReaders object so the user knows if its Cartesian or Spherical data
- dual return mode data from a Horizon (and Tele-15) sensor records all single returns as well as dual returns per laser pulse
- returned point data units are either meters or decimal degrees
- point data includes return number attribute for every point (constant 1 for single return data types should also be included)
- standard order for returned point data (Cartesian and Spherical) - [X, Y, Z, reflectivity, returnNum] or [Distance, Zenith, Azimuth, reflectivity, returnNum]
- if sensor uses 1 byte ’tag information’ its integer equivalent is recorded after returnNum
- support for multi-return dataset from MID-40/100 via the special firmwares

anjalizope commented:
I updated BinaryFileReader.py with the modifications above
